rule/unused-receiver.go

```diff
@@ -0,0 +1,77 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// UnusedReceiverRule lints unused params in functions.
+type UnusedReceiverRule struct{}
+
+// Apply applies the rule to given file.
+func (_ *UnusedReceiverRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	w := lintUnusedReceiverRule{onFailure: onFailure}
+
+	ast.Walk(w, file.AST)
+
+	return failures
+}
+
+// Name returns the rule name.
+func (_ *UnusedReceiverRule) Name() string {
+	return "unused-receiver"
+}
+
+type lintUnusedReceiverRule struct {
+	onFailure func(lint.Failure)
+}
+
+func (w lintUnusedReceiverRule) Visit(node ast.Node) ast.Visitor {
+	switch n := node.(type) {
+	case *ast.FuncDecl:
+		if n.Recv == nil {
+			return nil // skip this func decl, not a method
+		}
+
+		rec := n.Recv.List[0] // safe to access only the first (unique) element of the list
+		if len(rec.Names) < 1 {
+			return nil // the receiver is anonymous: func (aType) Foo(...) ...
+		}
+
+		recID := rec.Names[0]
+		if recID.Name == "_" {
+			return nil // the receiver is already named _
+		}
+
+		// inspect the func body looking for references to the receiver id
+		fselect := func(n ast.Node) bool {
```

Comment:
We should handle local variables which are named after the receiver

Reply:
Using the [`Obj`](https://golang.org/pkg/go/ast/#Object) field of the receiver identifier to search references to it guarantees us no collision with other identifiers.